### ISSUE TITLE
Conversation: MPC Gradient Marketplace

### ISSUE THREAD

**Issue by iamtrask** [OpenMined/OpenMined#27]
Over the course of this conversation, we'd like to take the eventual goal of a Multi-Party Compute Gradient Marketplace and break it into bite sized pieces we can work on. I'd like for us to kick that conversation off here.

Project 1: Simple gradient Tree Smart Contract - where people can submit gradients and pick their parent.

**Comment by diegoalejogm** [comment]
I'll pick this up

**Comment by iamtrask** [comment]
Note - for this first piece - if we get the simple Gradient Tree working with decrypted gradients - that'll be a nice starting point for working our way towards one compatible with MPC or HE.

**Comment by iamtrask** [comment]
For those watching at home - there's been a big offline conversation around how renumeration should be happening. There's lots of things still in the air - but a few things have become more solid. First, people should be able to pick which previous gradient to improve upon. In other words, there should be a tree where the Root of the tree is the randomly initialized model, and gradients are branches.

**Comment by emregoynugur** [comment]
Hey,

I am reposting my question from Slack: 

Can people submit as many gradients as they want to a particular model? e.g. if a model requests some personal info like gender, age etc., can people still compute several gradients? This would mean that they are actually processing the information of another person, right? Is this something permitted/desired?

**Comment by diegoalejogm** [comment]
I don't understand how people would be processing the information of another person. Each gradient's data is encrypted, so other's can't obtain the data from existing gradients. The posible issue is the quadratic growth of the trees, if miners try to mine starting on every node.

**Comment by emregoynugur** [comment]
Sorry, if it wasn't clear. I was wondering if a scenario like below would happen? (if it is allowed/prohibited/encouraged)

- Data Scientist asks for personal information of users
- User A accepts the request and submits his Facebook profile (age, location, gender etc.)
- User A gets paid for his contribution (so far so good)
- User A logs in with **a different** Facebook account and submits its profile
- User A gets paid **again**?

**Comment by diegoalejogm** [comment]
That is a scenario we could handle somehow... Maybe by checking duplicate hashes before training... Anyways re-training multiple times would reduce the gradient the same sample gives, so in the limit that person would stop earning any significative money. Moreover, the model could start overfitting and on that point that user certainly won't earn any.

Regarding the tree I am creating, there are currently no generics on Solidity, so the data-structure is merged with our structs in this iteration. We shall be checking  https://github.com/ethereum/solidity/issues/869 for any updates.

**Comment by emregoynugur** [comment]
I think you misunderstood me. I was just asking if it is possible to submit fake data if it improves the performance. I am particularly referring to information that maps one-to-one like age, height, gender etc.

For instance, after submitting her own profile, could a user create a new Facebook profile and submit another gradient, which could yield better performance if the sample is properly created.

I am aware that preventing fake data is very difficult, but for instance we know that certain information can only be submitted once.  I was hoping that being able to limit the number of gradients submitted by a person could discourage submitting fake data. 

Data scientists could determine the number of gradients submitted per user. If they only care about the performance (not fake data), they could allow users to submit as much as they want.

**Comment by diegoalejogm** [comment]
Oh I understand your point now. Sorry for the misunderstanding. 

That's true, limiting the number of gradients submitted by one person could discourage that. Though people could still create their first gradients with fake data.

What could be useful then is limiting the number of gradients a user submits to the same branch in the tree.

**Comment by iamtrask** [comment]
@erfannoury I answered this question here -https://www.youtube.com/watch?v=47enlQQcMQc

**Comment by iamtrask** [comment]
Basically - there's no way to limit the number of gradients that a person can submit - because everyone is anonymous. However, if someone submits the same gradient over and over, they'll cause the model to overfit. Branching explicitly allows people to overcome this by skipping gradients that appear to be overfitting.

**Comment by andorsk** [comment]
Assuming you get enough scale it'll be pretty hard for any single person to contribute enough toward a model that it starts overfitting toward their own data. @iamtrask if they are overfitting the model the accuracy will go down on the test set (assuming their test set isn't also biased). Why does it matter if they submit multiple gradients? As long as the accuracy improves it shouldn't matter, and if the data scientists chose a bad algorithm for accuracy measurement (or a bad test set), then that should be his problem.

**Comment by jpgbus** [comment]
> For those watching at home - there's been a big offline conversation around how renumeration should be happening. 

Instead of focusing on how one's profile components will affect the magnitude of a gradient, how about systematically rewarding accuracy and completeness of one's profile?

In every sense of the term, data integrity is paramount. Properly structuring this incentive such that folks are willing to give the same level of effort and care in getting it right that they would when submitting personal info to, say, government agencies or medical care providers, where there are very clear incentives to not get the data wrong, imo, is the grand challenge. 

I say that financial compensation should be uniformly distributed to all members of a given data set. That someone matches the filter criteria for a given data set should not be their concern if the process is truly anonymous. That they have as much of their profile complete and available for which a given filter can be applied would be a natural incentive, as their compensation would scale accordingly. 

Perhaps this conversation is better served by considering how a person's profile, a [composite of components collected across chains](https://interledger.org), can be simply procured, amenable for given OM use cases, while making it as simple for end users to opt in or out of any OM use case for which they qualify.
